controllers/managedfusion_controller.go

```diff
@@ -297,26 +298,39 @@ func (r *ManagedFusionReconciler) initReconciler(ctx context.Context, req ctrl.R
 
 func (r *ManagedFusionReconciler) reconcilePhases() (reconcile.Result, error) {
 	if !r.managedFusionSecret.DeletionTimestamp.IsZero() {
-		if r.verifyOfferringsDoNotExist() {
-			r.Log.Info(fmt.Sprintf("removing finalizer from %s secret", managedFusionSecretName))
-			r.managedFusionSecret.SetFinalizers(utils.Remove(r.managedFusionSecret.GetFinalizers(), managedFusionFinalizer))
-			if err := r.Client.Update(r.ctx, r.managedFusionSecret); err != nil {
-				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from %s secret: %v", managedFusionSecretName, err)
+		r.Log.Info("Managed Fusion configuration Secret is marked for deletion, checking for ManagedFusionOfferings CRs on the cluster")
+		if found, err := r.checkForManagedOfferings(); err != nil {
+			return ctrl.Result{}, err
+		} else if found {
+			r.Log.Info("ManagedFusionOffering CRs found, issuing delete")
+			offering := v1alpha1.ManagedFusionOffering{}
+			if err := r.deleteAllOf(&offering); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to issue delete for one or more ManagedFusionOffering CRs, %v", err)
+			} else {
+				r.Log.Info("ManagedFusionOffering CRs delete issued, requeuing the event")
+				return ctrl.Result{RequeueAfter: 10}, nil
+			}
+		} else {
+			r.Log.Info("No offering found, removing finalizer from %s secret", managedFusionSecretName)
+			if removed := utils.RemoveFinalizer(r.managedFusionSecret, managedFusionFinalizer); removed {
+				if err := r.update(r.managedFusionSecret); err != nil {
+					return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from %s secret: %v", managedFusionSecretName, err)
+				}
+				r.Log.Info("Removing finallizer from %s secret completed successfuly", managedFusionSecretName)
 			}
-			r.Log.Info("finallizer removed successfully")
 
 			if err := r.initiateAgentUninstallation(); err != nil {
 				return ctrl.Result{}, fmt.Errorf("failed to initiate agent uninstallation %v", err)
 			}
 		}
 
 	} else if r.managedFusionSecret.UID != "" {
-		if !utils.Contains(r.managedFusionSecret.GetFinalizers(), managedFusionFinalizer) {
+		if added := utils.AddFinalizer(r.managedFusionSecret, managedFusionFinalizer); added {
```

Comment:
Why can't we use the utils provided by controller-runtime instead of implementing our own?
https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.5/pkg/controller/controllerutil#AddFinalizer

Reply:
Because their finalizer utils do not return an indication if the new finalizer already exists or added.
We can change our utils to use a controllerutils implementation but that results in an unneeded iteration on the finalizer list which is unnecessary for such a simple operation